### DISCUSSION TITLE
Nginx settings on webserver not working, only frontpage index will load

### DISCUSSION THREAD

**Discussion by Sofiety** [phalcon/cphalcon#15302] [Q&A]
Hi guys!

So I set up a webserver with Nginx on jelastic (laravel), and I can't get it to work properly.
When going to the main url, the frontpage works, but any other links or domain just gives 404 not found. So basically the index.php rerouting is not working.

Running phalcon 3.45 and Php 7.3 on the webserver.

I tried the original nginx config, then the frontpage works. If I try the Nginx settings from the documentation the webserver does not start at all...

Here is the original config from the webserver (only frontpage and main url working):

`#user  nobody;
worker_processes  auto;
worker_rlimit_nofile 2048;

#load_module modules/ngx_http_modsecurity_module.so;

error_log  /var/log/nginx/error.log;
#error_log  logs/error.log  notice;
#error_log  logs/error.log  info;

#pid        logs/nginx.pid;

events {
    worker_connections  2048;
}

http {
    server_tokens off;
    include       mime.types;
    default_type  application/octet-stream;
    client_max_body_size 100m;

    log_format  main  '$remote_addr:$http_x_remote_port - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for" '
                      '"$host" sn="$server_name" '
                      'rt=$request_time '
                      'ua="$upstream_addr" us="$upstream_status" '
                      'ut="$upstream_response_time" ul="$upstream_response_length" '
                      'cs=$upstream_cache_status' ;

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    #keepalive_timeout  0;
    keepalive_timeout  65;

    gzip  on;
    gzip_types text/plain text/xml text/css application/xml application/xhtml+xml application/rss+xml application/javascript application/x-javascript;

    # Websocket support
    #upstream websocket {
    #    server 127.0.0.1:<PORT>;
    #}

    server {
        listen       80;
        listen       [::]:80;
        server_name  _;

        #charset koi8-r;

        #access_log  logs/host.access.log  main;
	    include /etc/nginx/aliases.conf;

        #location /ws {
        #    proxy_pass http://websocket;
        #    proxy_http_version 1.1;
        #    proxy_set_header Upgrade $http_upgrade;
        #    proxy_set_header Connection "Upgrade";
        #}
        if ($http_x_remote_port = '' ) {
            set $http_x_remote_port $remote_port;
        }

        #modsecurity on;
        #modsecurity_rules_file /etc/nginx/conf.d/modsecurity/modsec_includes.conf;

        location / {
            root   /var/www/webroot/ROOT/public;
            index  index.html index.htm index.php;
    	    location ~ \.php$ {
		location ~ /\. { deny all; access_log off; log_not_found off; }
		include /etc/nginx/fastcgi_params;
		fastcgi_pass unix:/var/run/php-fpm.socket;
		fastcgi_param SCRIPT_FILENAME $realpath_root$fastcgi_script_name;
		fastcgi_param DOCUMENT_ROOT $realpath_root;
		fastcgi_param PATH_INFO $fastcgi_script_name;
        fastcgi_index /index.php;
    	    }

        }
        index  index.php index.html index.htm;

        #error_page  404              /404.html;

        # redirect server error pages to the static page /50x.html
        #
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   html;
        }

        # proxy the PHP scripts to Apache listening on 127.0.0.1:80
        #
        #location ~ \.php$ {
        #    proxy_pass   http://127.0.0.1;
        #}
        location ~ ^/\.well-known {    root /var/www/webroot/ROOT;  allow all;}
        location ~ /\.well-known {       root /var/www/webroot;  allow all;}
	location ~ /\. { deny all; access_log off; log_not_found off; }
        # pass the PHP scripts to FastCGI server listening on 127.0.0.1:9000
        #
        #location ~ \.php$ {
        #    root           html;
        #    fastcgi_pass   127.0.0.1:9000;
        #    fastcgi_index  index.php;
        #    fastcgi_param  SCRIPT_FILENAME  /scripts$fastcgi_script_name;
        #    include        fastcgi_params;
        #}

        # deny access to .htaccess files, if Apache's document root
        # concurs with nginx's one
        #
        #location ~ /\.ht {
        #    deny  all;
        #}
	location ~ \.php$ {
		location ~ /\. { deny all; access_log off; log_not_found off; }
		include /etc/nginx/fastcgi_params;
		fastcgi_pass unix:/var/run/php-fpm.socket;
		fastcgi_param SCRIPT_FILENAME /var/www/webroot$fastcgi_script_name;
		fastcgi_param PATH_INFO $fastcgi_script_name;
		fastcgi_param DOCUMENT_ROOT /var/www/webroot;
      fastcgi_index /index.php;
        }

    }

include /etc/nginx/conf.d/*.conf;

}

`

Here is the server settings from the wiki. As mention, if I use this the webserver wont start/gives error.
https://docs.phalcon.io/3.4/en/webserver-setup
`server {
    # Port 80 will require Nginx to be started with root permissions
    # Depending on how you install Nginx to use port 80 you will need
    # to start the server with `sudo` ports about 1000 do not require
    # root privileges
    # listen      80;

    listen        8000;
    server_name   default;

    ##########################
    # In production require SSL
    # listen 443 ssl default_server;

    # ssl on;
    # ssl_session_timeout  5m;
    # ssl_protocols  SSLv2 SSLv3 TLSv1;
    # ssl_ciphers  ALL:!ADH:!EXPORT56:RC4+RSA:+HIGH:+MEDIUM:+LOW:+SSLv2:+EXP;
    # ssl_prefer_server_ciphers   on;

    # These locations depend on where you store your certs
    # ssl_certificate        /var/nginx/certs/default.cert;
    # ssl_certificate_key    /var/nginx/certs/default.key;
    ##########################

    # This is the folder that index.php is in
    root /var/www/default/public;
    index index.php index.html index.htm;

    charset utf-8;
    client_max_body_size 100M;
    fastcgi_read_timeout 1800;

    # Represents the root of the domain
    # http://localhost:8000/[index.php]
    location / {
        # Matches URLS `$_GET['_url']`
        try_files $uri $uri/ /index.php?_url=$uri&$args;
    }

    # When the HTTP request does not match the above
    # and the file ends in .php
    location ~ [^/]\.php(/|$) {
        # try_files $uri =404;

        # Ubuntu and PHP7.0-fpm in socket mode
        # This path is dependent on the version of PHP install
        fastcgi_pass  unix:/var/run/php/php7.0-fpm.sock;

        # Alternatively you use PHP-FPM in TCP mode (Required on Windows)
        # You will need to configure FPM to listen on a standard port
        # https://www.nginx.com/resources/wiki/start/topics/examples/phpfastcgionwindows/
        # fastcgi_pass  127.0.0.1:9000;

        fastcgi_index /index.php;

        include fastcgi_params;
        fastcgi_split_path_info ^(.+?\.php)(/.*)$;
        if (!-f $document_root$fastcgi_script_name) {
            return 404;
        }
        
        fastcgi_param PATH_INFO       $fastcgi_path_info;
        # fastcgi_param PATH_TRANSLATED $document_root$fastcgi_path_info;
        # and set php.ini cgi.fix_pathinfo=0
        
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
    }

    location ~ /\.ht {
        deny all;
    }

    location ~* \.(js|css|png|jpg|jpeg|gif|ico)$ {
        expires       max;
        log_not_found off;
        access_log    off;
    }
}`

How can I combine this to get it to work?

Thank you!!!!!

Best Regards,
Johan

**Comment by renatogabrielbr** [comment]
If i understood you problem you should set at least one with default_server or set server_name

```
listen       80;
listen       [::]:80;
**server_name *.wiki.com**;
```

And in the default you should do, all routes not matched the above server name will get in here.

```
listen 80 default_server;
listen [::]:80 default_server;
```


I use like this, it get the subdomain and matches with a folder, on the second if none was match I go to default,
After this conf I had put a another if example.

```
server {
	listen 80;
	listen 443;
	server_name *.dev.site.com;
	
    index index.php index.html;
	set $domain $host;

	client_max_body_size 512M;

        #subdomain match
	if ($domain ~ "^(api|admin|cms|).dev\.site\.com$") {
		set $domain $1;
		set $rootpath "site-${domain}/public";
	}

       #any subdomain match if none above
	if ($domain ~ "^(.[^.]*)+\.dev\.site\.com$") {
		set $domain $1;
		set $rootpath "site-frontend/public";
	}

    #root path www.site.dev.com will be /var/www/html/www/public
    root /var/www/html/$rootpath;
	
	access_log "/var/log/nginx/log.access.log";
	error_log "/var/log/nginx/log.error.log";

	include snippets/custom-error-pages.conf;
	
	location ~* ^[^\?\&]+\.(html|jpg|js|jpeg|json|gif|png|ico|css|zip|tgz|gz|rar|bz2|doc|xls|pdf|ppt|txt|tar|mid|midi|wav|bmp|rtf|js|svg|woff|ttf)$ {
		expires 2d;
 		add_header Cache-Control "public, no-transform";
		add_header 'Access-Control-Allow-Origin' '*';
		try_files $uri $uri/ =404;
	}
	
	location / {
		add_header 'Access-Control-Allow-Origin' '*';

		if ($request_method = 'OPTIONS') {
			add_header 'Access-Control-Allow-Origin' '*';
			add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, PUT, DELETE';
			add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization';
			add_header 'Access-Control-Max-Age' 1728000;
			add_header 'Access-Control-Allow-Credentials' 'true';
			add_header 'Content-Type' 'text/plain; charset=utf-8';
			add_header 'Content-Length' 0;
			return 204;
		}

		#include snippets/header-control.conf;


		try_files $uri $uri/ /index.php?$args;
	}
	
	location ~ \.php$ {
		include snippets/header-control.conf;
		
		try_files $uri =404;
		fastcgi_split_path_info ^(.+\.php)(/.+)$;
		fastcgi_pass unix:/var/run/php/php7.4-fpm.sock;
		fastcgi_index index.php;

		fastcgi_param   APP_ENV dev;
		
		fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
		fastcgi_param SCRIPT_NAME $fastcgi_script_name;
		include fastcgi_params;
    }
}

```

You can remove both if and use only it

```
	if ($domain ~ "^(.[^.]*)+\.dev\.site\.com$") {
		set $domain $1;
		set $rootpath "site-${domain}/public";
	}
```

And you can do it ( remove nginx default conf with default_server)

```
	listen 80 default_server;
	listen 443 default_server;
	server_name _;

       if ($domain ~ "^(.[^.]*)+\.site\.com$") {
		set $domain $1;
		set $rootpath "site-${domain}/public";
	}
	
```
